base/applications/mspaint/mouse.cpp

```diff
@@ -277,293 +508,374 @@ endPaintingL(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
-        case TOOL_FILL:
-        case TOOL_COLOR:
-        case TOOL_ZOOM:
-        case TOOL_BRUSH:
-        case TOOL_AIRBRUSH:
-        case TOOL_TEXT:
-            break;
+            draw(bLeftButton, x, y, FALSE);
+        }
+
+        if (pointSP == _countof(pointStack))
+            pointSP--;
+
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        imageModel.ResetToPrevious();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_ELLIPSE
+struct EllipseTool : ToolBase
+{
+    EllipseTool() : ToolBase(TOOL_ELLIPSE)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Ellp(m_hdc, start.x, start.y, x, y, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+        else
+            Ellp(m_hdc, start.x, start.y, x, y, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_RRECT
+struct RRectTool : ToolBase
+{
+    RRectTool() : ToolBase(TOOL_RRECT)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            RRect(m_hdc, start.x, start.y, x, y, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+        else
+            RRect(m_hdc, start.x, start.y, x, y, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        draw(bLeftButton, x, y);
```

Comment:
Missing handling of `GetAsyncKeyState(VK_SHIFT)` + `regularize()` call.

Reply:
Fixed.